crates/services/parallel-executor/src/scheduler.rs

```diff
@@ -0,0 +1,1179 @@
+                        break 'outer;
+                    }
+                }
+            } else {
+                tokio::select! {
+                    _ = new_tx_notifier.notified() => {
+                        self.new_executable_transactions();
+                    }
+                    result = self.current_execution_tasks.select_next_some() => {
+                        match result {
+                            Ok(res) => {
+                                match res {
+                                    Ok(res) => {
+                                        if !res.skipped_tx.is_empty() {
+                                    self.sequential_fallback(block_height, &consensus_parameters, res.batch_id, res.txs, res.coins_used, res.coins_created).await?;
+                                    continue;
+                                }
+                                self.register_execution_result(res);
+                                    },
+                                    Err(e) => {
+                                        return Err(SchedulerError::ExecutionError(e));
+                                    }
+                                }
+
+                            }
+                            _ => {
+                                return Err(SchedulerError::InternalError(
+                                    "Worker execution failed".to_string(),
+                                ));
+                            }
+                        }
+                    }
+                    _ = tokio::time::sleep_until(deadline) => {
+                        break 'outer;
+                    }
+                }
+            }
+        }
+
+        self.wait_all_execution_tasks(
+            block_height,
+            &consensus_parameters,
+            block_constraints.total_execution_time,
+        )
+        .await?;
+
+        let mut res = self.verify_coherency_and_merge_results(
+            nb_batch_created,
+            components.header_to_produce,
+            l1_execution_data,
+        )?;
+
+        if self.blob_transactions.is_empty() {
+            let mut merged: Changes = Changes::default();
+            for changes in res.changes.extract_list_of_changes() {
+                merged.extend(changes);
+            }
+            let storage_with_res = self
+                .storage
+                .latest_view()
+                .map_err(SchedulerError::StorageError)?
+                .into_transaction()
+                .with_changes(merged);
+            let (blob_execution_data, blob_txs) = self
+                .execute_blob_transactions(
+                    components,
+                    storage_with_res,
+                    nb_transactions,
+                    consensus_parameters_version,
+                )
+                .await?;
+            res.add_blob_execution_data(blob_execution_data, blob_txs);
+        }
+
+        self.reset();
```

Comment:
Maybe it will be better if `Scheduler` were created per block. So in this case you don't need rest functionality and you can just consume `self` instead of `&mut self`.